dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/boogie/VisitorBoogie.java

```diff
@@ -486,6 +479,11 @@ public Object visitAssume_cmd(Assume_cmdContext ctx) {
 			} else {
 				pairingLabel = pairLabels.get(currentLabel);
 			}
+			// "assume false" simple terminates the thread
+			// pairingLabel is guaranteed to be "END_OF_T"
+			if(ctx.proposition().expr().getText().equals("false")) {
+				programBuilder.addChild(threadCount, EventFactory.newGoto(pairingLabel));
+			}
```

Comment:
Is this some kind of optimization for `assume(false)`? `assume(false)` does not simply terminate a thread, it is stronger than that. If you want to keep the assume semantics but also allow for DCE, I think you have to keep the assume and add the terminating goto afterwards (this goto will never get executed, but it makes sure that DCE can eliminate the subsequent code).

Reply:
Yes, it was intended to be an optimization. I think this code is quite old (before all our discussion about assume semantics). I agree it does not match our current status of assume semantics. I don't fully understand the second part of your comment. Do you mean adding an `Assume` event just before the `Goto` in line 485?

Reply:
Yes. The `Assume` event will give the intended semantics. The `goto END_OF_T` will just be there so that our control-flow analysis can understand that the subsequent code is actually dead (we could also extend our cf-analysis to reason about `Assume` but that seems a lot harder).

Reply:
There seems to be a mismatch of what we expect of the boogie code `assume(false)` and what smack generates.
The problem with the CI is that the original C code [calls `pthread_exit()`](https://github.com/sosy-lab/sv-benchmarks/blob/2e1723fde6aa65a250dcb677efa45edaa4b6b631/c/pthread/fib_bench-2.c#L32) which should terminate only the calling thread [according to the documentation](https://www.ibm.com/docs/en/i/7.2?topic=ssw_ibm_i_72/apis/users_18.html).

The code generated by smack looks like this
```
t2
(88)     skip
(89)     start_thread()
(90)     5:$p0 <- 0
(92)     5:$i1 <- 0
(94)  5:$bb1:
(95)     5:$i2 <- 5:$i1 < 5
(98)     if((! 5:$i2 == 1)); then goto 5:$bb3:
(99)     begin_atomic()
(100)    5:$i3 = load(*i)
(101)    5:$i4 = load(*j)
(102)    5:$i5 <- (5:$i4 + 5:$i3)
(103)    store(*j, 5:$i5)
(104)    end_atomic()
(107)    5:$i6 <- (5:$i1 + 1)
(108)    5:$i1 <- 5:$i6
(109)    goto 5:$bb1:
(110) 5:$bb3:
(111)    if((! (! 5:$i2 == 1))); then goto END_OF_T3:
(112)    assume(False)
(114) END_OF_T3:
```
which will execute the assume event 112 when the bound is large enough and thus blocking the execution which should fail.

Reply:
From [here](https://github.com/smackers/smack/blob/c7d0694f08cefb422ebcc67c23824727b06b370e/lib/smack/SmackInstGenerator.cpp#L329-L333) it is clear that smack converts any unreachable instruction into an `assume(false)` and thus it differs from our semantics for assume events